### DISCUSSION TITLE
Reverse proxy has trouble if not at root level

### DISCUSSION THREAD

**Discussion by lpyparmentier** [firefly-iii/firefly-iii#4534] [Firefly III]
**Bug description**
I am running Firefly III version jc5x/firefly-iii:version-5.4.4 , and my problem is:

Similar as #2848 or #4467. Firefly run under a container, which is cool to host on arm board at home, however if the reverse proxy is not set-up as  root location (`location /`), something goes wrong with the redirection. 

Why it matters ? the benefit of arm board with container is to have multiple services on the same board for self-hosting stuff. It kills all the benefit here. If other services bring the same issues, I can't deal with multiple services on the same host. I have to change the code for each project and build dedicated image. I'm saying that because based on past issues it's always closed and considered as isolated case or not a problem related with firefly which in fact is. 

**Steps to reproduce**
run jc5x/firefly-iii:version-5.4.4 container and set-up a reverse proxy with location /firefly 

Some propose a hack like in #4243, if i have to hack each project it becomes a waste of time... Can't do that for each of the services I host at home :)

EDIT: also tried https://github.com/firefly-iii/firefly-iii/issues/1869#issuecomment-450594411, got `FireflyIII 404 Not Found :(`.

**Comment by JC5** [comment]
It's not a problem with Firefly III, and I will close this ticket just the same. You have three options.

1. Use subdomains if you want to host multiple services on one host.
2. Find a reverse proxy that will not just change `host:port` but `host:port/path`.
3. Build a Docker image in which the webserver believes it's being served at `/whatever` so it matches up with your reverse proxy config.

The Firefly III Docker image has a built in Apache that assumes it's at `/`. Reverse proxy's can't change that so if you start aliassing stuff to `/firefly-iii` they will forward the `/firefly-iii` part to Firefly III who will then give a 404.

I built the Docker image for the `/` use case and it services 99% of people using Firefly III. If you have to hack every project to get it your way maybe this isn't my problem.

**Reply by JC5** [reply]
I just realised we now have discussions so nothing gets closed anymore.

**Reply by lpyparmentier** [reply]
That's good for a debate :)

> 1. Use subdomains if you want to host multiple services on one host.

Even If I would, it does not work in my case. I have a **main** board acting like a reverse proxy to dispatch the external traffic to different internal services (not necessarily on the same board) (e.g., `proxy_pass https://private-ip-with-regrouped-services/firefly/;`). And then, a **second** board hosting container services for example, which explains why I need a second reverse proxy to point on the internal network related to the container (.e.g., docker). 

> 2. Find a reverse proxy that will not just change host:port but host:port/path.

What do you mean ? that nginx can't do that ? 

> 3. Build a Docker image in which the webserver believes it's being served at /whatever so it matches up with your reverse proxy config.

This would be my last solution. I've 10 different self hosted services. Maintaining them take time. If i start to change the code for each of them, I waste my time. Let say a year passed, and you realized that you need to update the image (security reasons, features), some variables and standard of the project changed. So you need to spend again your time to make your hack works again. Yes it's a quick fix, but on long term it's not the way to go. 

> I built the Docker image for the / use case and it services 99% of people using Firefly III. If you have to hack every project to get it your way maybe this isn't my problem.

Don't know where you find this exact 99% percent ? However a lot of other services are capable of working with such a configuration of reverse proxy (with no hack), just through configuration. 

EDIT: In fact it's pretty standard to do that. The containers/VM change the standard in a bit, while in the past we hosted one service on a single device, now we easily host multiple services on the same host which is fine for a self-host service where you have 1 or 2 users. **Problem with that is the fact that "subfolder" become a useful part in the reverse proxy process.**

**Reply by JC5** [reply]
> Even If I would, it does not work in my case.

Yes it would. I use the same setup. Here's (part of) an example nginx config. I proxy pass over different ports to differentiate the services on the same host if necessary.

```
# https
server {
	# mod security
	modsecurity on;
	modsecurity_rules_file /etc/nginx/modsec/main.conf;

	listen 443 ssl http2;
	listen [::]:443 ssl http2;
	root /var/www/html;
	index index.html;

	server_name subdomain.example.com;

	location / {
		proxy_set_header X-Forwarded-Host $host;
		proxy_set_header X-Forwarded-Server $host;
		proxy_set_header X-Forwarded-Proto $scheme;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header Host $host;
		client_max_body_size 64M;
		proxy_read_timeout 300s;

		# pass to Docker container
		proxy_pass http://172.16.x.x:10011;
```

> What do you mean ? that nginx can't do that ?

Nginx can reliabily rewrite url's, but I haven't been able to reliably tell the underlying Docker container (through headers or something) that the virtual path is a subdirectory.

> Don't know where you find this exact 99% percent ? 

99% is my guess. The Docker image for Firefly III sits at 10+ million pulls and I have very few issues related to this subject.

> However a lot of other services are capable of working with such a configuration of reverse proxy (with no hack), just through configuration.

As far as I have ever been able to find out, the only method that works *is* a hack: by forcing Firefly III to rewrite all it's URL's independent of what the web server says they should be.

**Reply by JC5** [reply]
In your example, your reverse proxy has (presumably) a subdirectory for Firefly III, but the proxy passes to another subdirectory? How does that even work? Do you proxy to another proxy that passes it to Firefly III?

**Reply by lpyparmentier** [reply]
> Yes it would. I use the same setup. Here's (part of) an example nginx config. I proxy pass over different ports to differentiate the services on the same host if necessary.

You don't have the same set-up. You only have one reverse proxy. It's because your reverse proxy, your container, (and the public ip) is all on the same machine. Am I wrong ? 

> 99% is my guess. The Docker image for Firefly III sits at 10+ million pulls and I have very few issues related to this subject.

More popular will be your project, more people will face this kind of issue, and maybe give up to stay on root. It's problematic if all other services force you to root. I've no doubt your project will be successful, it's pretty cool and much advanced compare to the other. BTW I started to handle my cryptocurrencies with your project, and some currencies are 18 decimals while firefly limit to 12 max (it's another issue, but not important now).

> As far as I have ever been able to find out, the only method that works is a hack: by forcing Firefly III to rewrite all it's URL's independent of what the web server says they should be.

That's weird, I should take time to investigate the code. BTW, i've tried the fix proposed by https://github.com/firefly-iii/firefly-iii/pull/2514 and it gave me a blank page.

> In your example, your reverse proxy has (presumably) a subdirectory for Firefly III, but the proxy passes to another subdirectory? How does that even work? Do you proxy to another proxy that passes it to Firefly III?

I think the best answer is my configuration :smile: 

1. main board with public interface (dispatcher) - handle any https traffic with a * certifaddicate for example, then dispatch to board hosted in my internal network:
```
server {
    listen 443 ssl;

    server_name firefly.domain.tld;

    ssl_certificate /root/webservice/cert.pem;
    ssl_certificate_key /root/webservice/key.pem;

    location / {
      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto http;
      proxy_pass http://192.168.1.x/firefly/;  # http://192.168.1.x:8080 (see edit below)
      proxy_read_timeout 90;
    }
}
```

**Note**:
 - `192.168.1.x/firefly/` subfolder useful since there are multiple services on `192.168.1.x`
 - `192.168.1.x` is a private ip on my network, unrelated with docker network
 - the board only intend to manage the multi-domains and http(s) traffic. Don't want to run a docker process that requires root user. It helps me to keep a clean set-up.

2. `192.168.1.x` a board with services (particularity interesting thanks to VM and containers):

```
# simplified without ssl for the example (it's private network it's okay)
server {
        listen 80 default_server;
        listen [::]:80 default_server;

    # firefly service (could certainly remove this reverse with a hack, see edit below)
    location /firefly {
      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto http;
      proxy_pass http://172.17.0.2/;
      proxy_read_timeout 90;
    }

    # gitlab service (could certainly remove this reverse with a hack, see edit below)
    location /gitlab {
      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto http;
      proxy_pass http://172.17.0.3/;  # http://192.168.1.x:8090 (see edit below)
      proxy_read_timeout 90;
    }

    # and so on
}
```
**Note**: need reverse proxy due to the internal network managed by docker : `172.17.0.x`.


EDIT: I certainly could solve that by running each service (firefly, gitlab) and expose them directly with `docker --network=host ...` with different port, but to my point of view it's a hack.

EDIT2: Just tried the hack with `--network=host` it works fine! The bad part of this solution is that I exposed the whole container service to my internal network. To my case is okay, but for other that want a control with a reverse proxy (e.g., require https internally) it could be an issue. I also do not benefit of the features provided by the docker network (if there are some).

**Reply by JC5** [reply]
> You don't have the same set-up. You only have one reverse proxy. It's because your reverse proxy, your container, (and the public ip) is all on the same machine. Am I wrong ?

Yes, but no matter. Your setup is indeed entirely different from mine.

> More popular will be your project, more people will face this kind of issue, and maybe give up to stay on root. It's problematic if all other services force you to root. I've no doubt your project will be successful, it's pretty cool and much advanced compare to the other. 

Thanks for the compliments. But keep in mind I'm not in this game to make people happy. If it doesn't work because what I offer you doesn't work for you well... you get what you pay for and everything is free. 

All the technology is open source and this use case is honestly very rare. I love to help and support you but I don't see it as my duty to change what I offer.

> BTW I started to handle my cryptocurrencies with your project, and some currencies are 18 decimals while firefly limit to 12 max (it's another issue, but not important now).

I'll write it down for future releases!

> EDIT2: Just tried the hack with --network=host it works fine! The bad part of this solution is that I exposed the whole container service to my internal network. To my case is okay, but for other that want a control with a reverse proxy (e.g., require https internally) it could be an issue. I also do not benefit of the features provided by the docker network (if there are some).

Why not map internal Docker container ports to system ports, and use ports in nginx? You can map `11222:8080` in the case of Firefly III, and in nginx there is nothing wrong with `Listen 11222` combined with a https certificate.

**Reply by lpyparmentier** [reply]
> Thanks for the compliments. But keep in mind I'm not in this game to make people happy. If it doesn't work because what I offer you doesn't work for you well... you get what you pay for and everything is free.

Can't agree more

> I'll write it down for future releases!

:+1: 

> Why not map internal Docker container ports to system ports, and use ports in nginx? You can map 11222:8080 in the case of Firefly III, and in nginx there is nothing wrong with Listen 11222 combined with a https certificate.

Just tried, when I go to `https://hosted-firefly:11222/`, it redirects to `https://hosted-firefly`

EDIT: however it works if I go through the subdomain.

**Reply by JC5** [reply]
Mmmm. Assuming Firefly III is on another device.

- Expose it on that device only using `11222:8080`
- Reverse proxy it from that device using something like `Listen 9000; proxy_pass localhost:11222; ssl_settings [..]`
- On the main device, use `proxy_pass that_device:9000`

This needs a double set of X-forward header stuff, so the host is preserved twice.

**Reply by lpyparmentier** [reply]
My bad it works in both case. Can't explain why it didn't just few minutes ago. Case solved I guess! Thanks a lot for your time ! Hope it will help other in similar situation :smile: 

EDIT: In fact the page show up if i go `https://that_device:4430/login` but can't connect. And if i go `https://that_device:4430` it redirects to `https://that_device/` (not the right port)

Any way, it's fine for now, I'll go with the subdomain solution

**Reply by JC5** [reply]
Another alternative would be to set up some kind of VPN / virtual network between the devices.

**Reply by risacher** [reply]
I run Wekan, ownCloud, Wordpress, ttyx, etherpad, Cockpit, noVNC, Apache-httpd and about 6 more custom apps from the same FQDN, each in subdirectories (except httpd, which handles '/'). A reverse proxy unifies those apps into a single namespace.  All these apps either strictly use relative URLs or have a BASE_URL variable somewhere.  I'm enthused about adding Firefly, but lack of reverse-proxy-friendliness has me tripped up for the moment. 

I read your distaste for BASE_URL in the thread for #2514, and I'm not sure I understand why you object to it.  If someone were to revisit #2514, and find the edge cases that caused @TimCinel to give up, would you consider accepting it?

**Reply by JC5** [reply]
Just to be clear: Firefly III has zero problems with subdirectories. It runs 100% fine in `/whatever`.

Right now, as not to drag up issues and tech from 2 years ago, the situation is that you can't make the **standard** Docker image accept `/path` without changing the configuration of the Docker image's internal Apache server to pretend to be running at `/path`. Building your own Docker image to match `/path` is currently the only solution.

I still believe overruling your app to use a standard BASE_URL is a stupid hack, and I don't care that all the other kids are doing it too.

Although any reverse proxy worth it's salt is capable of communicating the host to the Docker image with zero configuration necessary, it's (still) not possible to do any translation between `/` (assumed by the Docker image) and `/path`, assumed by the reverse proxy. nginx can't do it, Caddy can't do it, none of them seem capable of doing this. They all suggest that you give the container knowledge it shouldn't need to have, ie. the path the reverse proxy is running it under.

But at some point something's got to give and Laravel has changed a lot over time so allow me to look into an elegant solution for this, because I'm sure it's out there.

**Reply by Flightkick** [reply]
@JC5 
> none of them seem capable of doing this. They all suggest that you give the container knowledge it shouldn't need to have, ie. the path the reverse proxy is running it under.

Reverse proxies generally do not rewrite content of the resources they serve (in this case, the HTML page). The proxy pass configurations support routing `/` to `/path` and rewriting `Location` headers for redirects but that's about where the capabilities end. (unless using some mods like [http_sub_module](https://nginx.org/en/docs/http/ngx_http_sub_module.html), but talking about silly hacks).

I agree that passing a base path does not look like a clean solution. It'd be much nicer if the reverse proxy and the routing middleware of Laravel would automatically resolve this. Unfortunately they can't. The official [rfc7239 ](https://datatracker.ietf.org/doc/html/rfc7239) does not include support for an Uri or Path header which might have allowed the proxy to communicate the base url to Laravel without any user configuration. Laravel, more specifically TrustedProxy do not offer support for any unofficial headers either.

As an alternative, making the application entirely agnostic of it's context would require all internal urls to be relative. This means you'd need to remove  `<base href="{{ route('index') }}/">` from the templates, although I'm not sure if there'd be consequences in other parts of the application. After that, you'd need to find a way for Laravel to generate only relative urls. I'm not familliar with Laravel but it seems like the `Route::get()` approach is preferred instead of `URL.to()` which is currently being used (the latter seems to generate absolute urls).

Anyways, with the current setup I don't really get why you'd be so much against a base url config at this point in time. The application already has this context information through the `APP_URL` (which is currently only used for email generation), but the information is there already. A lot of the templates are including the HTML `<base>` tag and the application is not ready to fully depend on relative urls. Why not leverage the framework's functionality and instruct Laravels middleware to generate it's URLs aware of it's current context.

For the argument of not applying 'stupid hacks'. Building and maintaining a Docker image for the sake of using a subpath seems like a lot more waste. A subdomain unfortunately has other consequences which are not always preferred.

I hope you might reconsider the proposals to allow a configurable base path.
Nevertheless, thank you very much for your time and effort 🙂 

PS: Apparently due to a lazy search I missed this Q&A and https://github.com/firefly-iii/firefly-iii/pull/2514 so at the time I created my own https://github.com/firefly-iii/firefly-iii/issues/4933 which more or less does the same thing. I've "tested" my implementation with my proxy configuration and did not stumble upon any issues. I especially checked creating and deleting transactions as these issues have been mentioned in the other PR. So far no issues found.

**Reply by JC5** [reply]
URL.to is used in a few templates but regardless; i have to change everything or nothing. If there's going to be a hack to change the URL (as opposed to making all links relative which brings its own problems), I'd like to make the path configurable, because the host/domain doesn't need it. That needs its own rewrites but not as excessive.

**Reply by Flightkick** [reply]
In all fairness, I do not think that 'hack' is a proper term for supporting configurable base paths. If anything, injection of such properties decouples the behavior from infrastructure which is arguably a good thing.

There might be a way to allow a configurable base path without overriding the scheme and/or host. Although that would probably require some refactoring of routing and might bring challenges in other area's. Leaving you with a custom implementation instead of being able to delegate that part entirely to the framework.

Given that the `APP_URL` already needs to be present in the config for building email links, the ROI in terms of time/benefit and the minimal impact on code quality this really looks like a no-brainer issue. Respectfully, rather overhauling the entire thing instead of accepting a slight trade-off from Laravel's design decisions sounds like over-engineering to me.

**Comment by hjhee** [comment]
Thanks for the information. I would like to backup and share my working configuration by adding an ngingx reverse proxy server to the docker-compose.yml, no need to build another docker image like in option 3 above.

I adapted [#1869](https://github.com/firefly-iii/firefly-iii/issues/1869#issuecomment-450594411) and got the following section in nginx config file:
```
    location /fireflyiii/ {
        proxy_pass http://app:8080/;
        proxy_pass_request_headers on;
        proxy_set_header Host $host;
        proxy_set_header Accept-Encoding "";

        # Rewrite redirects, such as / to /login and /logout
        # No trailing slash - including it works for login, but (at least) not logout.
        # Also means we can't use `proxy_redirect default;`
        proxy_redirect $scheme://$host /fireflyiii;

        # Setting proper X-Forwarded headers, in combination TRUSTED_PROXIES
        # env variable, lets Laraval create URLs based on proxied address rather
        # than server name.
        proxy_set_header X-Forwarded-For $remote_addr;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Port $server_port;

        # Add the folder name to links/asset URLs in the response body
        sub_filter $scheme://$host https://$host/fireflyiii;
        sub_filter_once off;
        # sub_filter_types *;
    }
```

In my setting, I used traefik as the main reverse proxy, but I'm more familiar with nginx and it is flexible enough to achieve what I want.
docker-compose.yml
```yaml
version: '3.3'

services:
  app:
    image: fireflyiii/core:latest
    restart: always
    volumes:
      - firefly_iii_upload:/var/www/html/storage/upload
      # - ./000-default.conf:/etc/apache2/sites-available/000-default.conf:ro
    env_file: .env
    networks:
      - default
      - xwiki-postgres
      # - traefik-gateway
  nginx:
    image: nginx:mainline
    restart: "always"
    labels:
      - traefik.enable=true
      - "traefik.http.routers.fireflyiii.entrypoints=websecure"
      - "traefik.http.routers.fireflyiii.rule=Host(`ubuntu-e480`) && PathPrefix(`/fireflyiii`)"
      - "traefik.http.services.fireflyiii.loadbalancer.server.port=80"
      - traefik.http.routers.fireflyiii.service=fireflyiii
      - traefik.http.routers.fireflyiii.tls=true
      - "traefik.docker.network=traefik-gateway"

      - "traefik.http.routers.fireflyiii-insecure.entrypoints=web"
      - "traefik.http.routers.fireflyiii-insecure.rule=Host(`ubuntu-e480`) && PathPrefix(`/fireflyiii`)"
      - "traefik.http.routers.fireflyiii-insecure.middlewares=fireflyiii-insecure-mw"
      - "traefik.http.middlewares.fireflyiii-insecure-mw.redirectscheme.scheme=https"
      - "traefik.http.middlewares.fireflyiii-insecure-mw.redirectscheme.permanent=false"
      - "traefik.http.routers.fireflyiii-insecure.service=noop@internal"
    depends_on:
      - app
    networks:
      - default
      - traefik-gateway
    volumes:
      - ./site.default:/etc/nginx/conf.d/default.conf:ro
volumes:
   firefly_iii_upload:
   firefly_iii_db:
networks: 
  traefik-gateway:
    external: true
  xwiki-postgres:
    name: xwiki_xwiki-postgres
    external: true
```

**Reply by Cloud9Developer** [reply]
The nginx config shared above works great for most things!

1 thing I noticed was I was unable to change the dates in the ui (was getting error messages because the /daterange & /transactions path was being called on the root path (obviously 😄 )), so had to add another location block to handle that (see below):

```
location /fireflyiii/ {
      	proxy_pass http://localhost:8038/;
      	proxy_pass_request_headers on;
      	proxy_set_header Host $host;
      	proxy_set_header Accept-Encoding "";

      	# Rewrite redirects, such as / to /login and /logout
      	# No trailing slash - including it works for login, but (at least) not logout.
      	# Also means we can't use `proxy_redirect default;`
      	proxy_redirect $scheme://$host /fireflyiii;

      	# Setting proper X-Forwarded headers, in combination TRUSTED_PROXIES
      	# env variable, lets Laraval create URLs based on proxied address rather
      	# than server name.
      	proxy_set_header X-Forwarded-For $remote_addr;
      	proxy_set_header X-Forwarded-Host $host;
      	proxy_set_header X-Forwarded-Proto $scheme;
      	proxy_set_header X-Forwarded-Port $server_port;
	      
	# Add the folder name to links/asset URLs in the response body
       sub_filter $scheme://$host https://$host/fireflyiii;
       sub_filter_once off;
      # sub_filter_types *;
}

location ~ ^/(daterange|transactions) {
	resolver 127.0.0.53;
	proxy_pass $scheme://$host/fireflyiii$request_uri;
}
```

I needed the resolver since I'm passing in variables to proxy_pass (was giving errors without it).

**Note, I am running nginx installed directly on my OS (Ubuntu 24.04) and Fireflyiii via docker**

There may be other URIs other than `daterange` and `transactions`, but those are the only 2 i've found so far. If you find others, you'll just need to add them in a similar fashion (unless someone smarter comes along and figures out how to dynamic capture all endpoints, while leaving the root path `/` empty 😄 )
